### DISCUSSION TITLE
RFC: useForm utilities props / improve existing clone

### DISCUSSION THREAD

**Discussion by bluebill1049** [react-hook-form/react-hook-form#8030] [RFC]
# Context

To avoid internal mutation, react hook form internally clone any external state which gets a pass into the hook. The original goal is to cover usage on standard primitive form value. However, in some applications users are try to store external custom Object, Set, or Map into the library. (A good example would be a moment custom object.) Those data types are currently not supported.

# Proposal

1. Introduce a config option that allows users to overwrite the internal clone method. This allows users to expand on other methods if the user has specific requirements such as data clone or even object comparison.

```tsx
// the following is custom clone function which you could supply
const cloneValues = (defaultValues) => {
    return Object.keys(defaultValues)
        .reduce((clone, key) => {
            const value = defaultValues[key];
            const replicator = replicators.find(value);
            return {
                ...clone,
                [key]: replicator ? replicator(value) : value,
            };
        }, {}};
};

const form = useForm({
  utilities: {
    clone: cloneValues,
  }
})
```

2. https://caniuse.com/?search=structuredClone is worth considering as well, but browser support is not great just yet. Maybe a v8 choice with polyfill at the app level.

3. check if the object is a plan object before cloning
https://github.com/jonschlinkert/is-plain-object

### Props
- Flexible
- Support custom data type

### Cons
- increase library complexity 
- increase bundle size

# Feature Request
https://github.com/react-hook-form/react-hook-form/issues/7279

# PR

- option 1: https://github.com/react-hook-form/react-hook-form/pull/8032
- option 3: https://github.com/react-hook-form/react-hook-form/pull/9080

cc @Moshyfawn @kotarella1110 @jorisre @barrymay @felixschorer

**Comment by Moshyfawn** [comment]
I see how that helps a lot with more complex data types and "unlocks" a better 3rd party data libs integration, per say. But I can't shake this thought that the problem here being solved is a problem about the lib internals that usually wouldn't be on the table for the users to vote on. What I mean by that is that thee issue of immutability isn't directly what RHF solves; it's about how the lib works and not the actual goal of the lib.

All of my blabling is to say, maybe, we can do a kind of RnD and see how different libs of different or the same goal solve the immutability? Maybe it's not even for debate what RHF doing cloning objects?
I don't remember, say, Redux Forms having a long ongoing discussion with their users about their internals and having a feature request to customize the internal cloning mechanism of theirs.

And I'm not trying to directly compare the experience different libraries have on the OSS field. If anything, I'm just laying my thoughts down to potentially spark some ideas about the topic :P

That said, feel free to completely dismiss or ignore the comment, if it doesn't make sense to you.

**Reply by bluebill1049** [reply]
Thanks for the input 🙏 

> feature request to customize the internal cloning mechanism of theirs.

I think that's exactly what we try to address here, the solution is a proposal at the moment. It may not be the right or correct solution, but this is indeed a problem to be addressed as more and more issues are related when users start to use form values other than primitive data types, array and objects. It's a similar issue where my initial proposal was to align web standards by unmount inputs and removing their value, but in real-world more and more apps would form libraries to also act as a "global" state manager, hence from v7 onward I have flipped the default with `shouldUnreigster: false`,  I think this issue shares some similarity, as most of the app would be fine to live without such, but it's difficult to keep it up with all those different data types, especially custom data object such as Luxon or moment, and I hope we don't have to dep on external lib either still keep this lib light and slim.

**Reply by bluebill1049** [reply]
The other option is to strictly state here the data type that we are not supporting, eg custom object.

**Reply by bluebill1049** [reply]
One more issue with the custom objects is a dirty flag for deep equal as well, which is another problem. which would generate from this approach. so users would have to supply their own clone and equal. maybe this is not a good solution after all. user should supply form field "value" instead of those non-native supported values.

**Comment by bluebill1049** [comment]
Going to pause this RFC, the more I think about it the less I like the API. I will document the supported data type in the documentation for now.

**Comment by nin-o** [comment]
An issue I'm currently having is: the instances of my class are being turned into plain objects. Maybe I could use this feature (not sure if there are any other workarounds currently).

For example:

```
import { useForm } from 'react-hook-form';

class Person {
    name: string = '';
}

export default function MyComponent() {
    const person = new Person();

    console.log(person instanceof Person); // true

    const {
        register, handleSubmit, watch
    } = useForm<Person>({
        defaultValues: person
    });

    const watched = watch();

    console.log(watched instanceof Person); // false

    function onSubmitted(submitted: Person) {
        console.log(submitted instanceof Person); // false
    }

    return (
        <form onSubmit={handleSubmit(onSubmitted)} noValidate>
            <input
                {...register('name')}
            />
            <button type="submit">Submit</button>
        </form>
    );
}
```

Thanks!

**Comment by scallaway** [comment]
This is something that I'm running into _constantly_ and am surprised that didn't have first-class support from the get-go.

The result of this is that I have to write a lot of logic to transform data into the shape that I expect it to be when pulling out values from `field` or `watch` etc. This pretty much removes most of the benefits of using a library like this in the first place.

I don't think it's an uncommon use-case to want to store a Set of strings or a Mapping of string to object in the Form state if you're looking to keep a collection of values as part of the input data.

I really do hope that this RFC gets picked up again and worked on as this is something I'm very keen on seeing in this library.

**Reply by kaisermann** [reply]
> This is something that I'm running into _constantly_ and am surprised that didn't have first-class support from the get-go.

Same here when dealing with values that are objects containing Firestore refs. While I could map them to strings and then map them back to the actual Ref object, I feel that I shouldn't need to be concerned with this. @bluebill1049 have you thought about a check that defines if some value should be cloned or not instead of customizing the cloning method? Something like [this](https://github.com/TehShrike/deepmerge#ismergeableobject)

**Reply by bluebill1049** [reply]
thanks for the feedback! Is plan object check seems to be a valid choice instead of utility method overwrite. let me update the RFC and see how it goes.

**Reply by bluebill1049** [reply]
here is the PR: https://github.com/react-hook-form/react-hook-form/pull/9080

**Comment by carlosbaraza** [comment]
It seems to me that having complex types is a really common thing for anyone integrating this library with a complex form. Developers integrating the form don't always control the data types of each component they use, so flexibility in this end is (in my humble opinion) *quite critical*.

In most major companies, there is a component library, usually developed by different teams and used by many, which makes it unfeasible to change the component data type in all the company verticals because RHF can't manage any kind of data type, while other teams might be using something else. Also, it is not clear at all what the user can do to deal with a complex data type (like DraftJS's EditorState, I opened #8214 to discuss this specific case). Is the user expected to serialise to string and deserialise again before using it to submit the form or validate?

There could be an interim solution, even if hacky, while a better approach is thought. Exposing the cloning internals seems a bit strange to me, because it is not clear to me as a user why is the data cloned in the first place. But, if this is the only way forward, it'd work, and at least provide a way out for anyone stuck with this (like me).

**Reply by bluebill1049** [reply]
Thanks for the feedback, your issue is different from the one above. In fact, it's already getting resolved in v8 (Type related): https://codesandbox.io/s/react-hook-form-handlesubmit-ts-v7-forked-r30kst?file=/src/index.tsx

**Comment by cjp-psl** [comment]
Is there any progress on this issue? This is a great library and it seems like an unnecessary limitation.

In my case I have a custom Date extension class which is currently being wiped out by the cloneObject implementation. It works fine when the value is changed (because my control returns the correct class) but if the value is not changed the class in submit data is wrong due to the clone.

Having the ability to provide a clone override, or even a simple callback to inject additional logic ahead of the current logic would solve a lot of potential issues (mine included), and doesn't *seem* like it is either too much work to implement or likely to cause downstream issues (though obviously there may be complexities I'm not seeing)

Happy to come up with a PR to implement this if it is a question of effort.

**Comment by Baterka** [comment]
Any updates on this?
